standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/SharedCache.java

```diff
@@ -309,12 +302,19 @@ public int getObjectSize(Class<?> clazz, Object obj) {
     private Map<String, SQLDefaultConstraint> defaultConstraintCache = new ConcurrentHashMap<>();
     private Map<String, SQLCheckConstraint> checkConstraintCache = new ConcurrentHashMap<>();
 
+    private boolean pk = false;
```

Comment:
Using too many flags is complex. Check if we can simplify it using below logic. 
1) Fetch all constraints and update cache -> If this update is successful, then set "isAllConstraintsSet=true"
2) If partially successful due to memory constraint, then set "isAllConstraintsSet=false" and set bit position of completed constraint "long constraintStatusBits". We can assign one bit per constraintType and generally in the order in which we update it. It means, if we encounter a bit which is unset, then rest of the bits are unset too.
3) We never refresh individual constraints. Always try to update all constraints by using same method which follow above logic to set the flag. Shall remove all individual update(CachedStore.updateTablePrimaryKeys...) and refresh(SharedCache.refreshPrimaryKeysInCache...) methods for each types. This ensure, we always cache a valid snapshot.
4) When read any constraint, first check isAllConstraintsSet==true. If true, then assume all are valid and return the cached entry immediately. If false, then just the corresponding bit in "constraintStatusBits" and if true, then return the cached entry or else return null and invoke RawStore.

Note: We can also get rid of isAllConstraintsSet flag by using constraintStatusBits = -1 for "true" scenario.

Reply:
Changes are as follows 
1. Replace 6 flag with 1 flag
2. In cache refresh/update flow removed 6 call to raw metastore with getAllTableConstraints()
3. While adding table to cache via createTable and createTableWithConstraint isConstraintValid flag is set as true.
4. If  isConstraintValid flag is true then return data from cache else redirect to raw metastore.
5. While adding individual constraint to metastore if  isConstraintValid is true that means constraints snapshot is valid and new constraint can be updated in cachestore else skip update.
6. isConstraintValid will only be true if all the constraint value are valid in the metastore else redirect to raw metastore (For the case of partial constraint storage due to memory shortage)